File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int[] inD = new int[numCourses];
        helper(inD, prerequisites);
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < inD.length; i++) {
            if (inD[i] == 0) {
                q.add(i);
            }
        }
        List<Integer> res = new ArrayList<>();
        while (!q.isEmpty()) {
            int curr = q.poll();
            res.add(curr);
            for (int i = 0; i < prerequisites.length; i++) {
                if (prerequisites[i][1] == curr) {
                    inD[prerequisites[i][0]]--;
                    if (inD[prerequisites[i][0]] == 0) {
                        q.add(prerequisites[i][0]);
                    }
                }
            }
        }
        if (res.size() != numCourses) {
            return new int[0];
        }
        int[] finRes = new int[res.size()];
        for (int i = 0; i < res.size(); i++) {
            finRes[i] = res.get(i);
        }
        return finRes;
    }

    public void helper(int[] inD, int[][] prerequisites) {
        for (int i = 0; i < prerequisites.length; i++) {
            inD[prerequisites[i][0]]++;
        }
    }
}

File: 0210-course-schedule-ii/0210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj = [[] for _ in range(numCourses)]
        in_degree = [0] * numCourses
        for course,pre in prerequisites:
            adj[pre].append(course)
            in_degree[course] += 1
        q = deque([i for i in range(numCourses) if in_degree[i] == 0])
        ans = []
        while q:
            curr = q.popleft()
            ans.append(curr)
            for i in adj[curr]:
                in_degree[i] -= 1
                if in_degree[i] == 0:
                    q.append(i)
        print(ans)
        return ans if len(ans) == numCourses else []